Add fused layer norm impl on CUDA in PyTorch

xiaomengy opened the pull request:
Summary:
Add fused layer norm impl on CUDA in PyTorch

The second or higher order gradient part is copied from https://github.com/pytorch/pytorch/pull/26201. Thanks @zasdfgbnm provide a good example for dealing with higher order gradients.

The benchmark results for this PR vs original LayerNorm as well as apex.FusedLayerNorm are shown below. We can see some improvements compare to apex.FusedLayerNorm especially on common batch sizes such as 128 or 256. 

Performance benchmark for original LayerNorm vs apex.FusedLayerNorm on a V100 machine
```
**************************************
Shape = (128, 2097152)
  curr LayerNorm forward: 11.19029480102472ms
  apex LayerNorm forward: 10.364733319962397ms
  curr LayerNorm backward: 27.58014868805185ms
  apex LayerNorm backward: 20.948871094966307ms
**************************************
Shape = (256, 1048576)
  curr LayerNorm forward: 10.13169369683601ms
  apex LayerNorm forward: 6.387616391992196ms
  curr LayerNorm backward: 26.80497354711406ms
  apex LayerNorm backward: 15.511555707082152ms
**************************************
Shape = (512, 524288)
  curr LayerNorm forward: 9.748662753030658ms
  apex LayerNorm forward: 4.730404008179903ms
  curr LayerNorm backward: 26.34211923298426ms
  apex LayerNorm backward: 14.006239669863135ms
**************************************
Shape = (1024, 262144)
  curr LayerNorm forward: 8.120901573915035ms
  apex LayerNorm forward: 5.376706059090793ms
  curr LayerNorm backward: 25.534225502051413ms
  apex LayerNorm backward: 14.285464284941554ms
**************************************
Shape = (2048, 131072)
  curr LayerNorm forward: 7.154732370981947ms
  apex LayerNorm forward: 4.762221570825204ms
  curr LayerNorm backward: 25.278658757917583ms
  apex LayerNorm backward: 13.5387117639184ms
**************************************
Shape = (4096, 65536)
  curr LayerNorm forward: 6.96491337986663ms
  apex LayerNorm forward: 4.47763676289469ms
  curr LayerNorm backward: 25.304558301810175ms
  apex LayerNorm backward: 13.59957629814744ms
**************************************
Shape = (8192, 32768)
  curr LayerNorm forward: 6.964353390969336ms
  apex LayerNorm forward: 4.3496931828558445ms
  curr LayerNorm backward: 25.336067016003653ms
  apex LayerNorm backward: 13.46561166504398ms
**************************************
Shape = (16384, 16384)
  curr LayerNorm forward: 6.937638100003824ms
  apex LayerNorm forward: 4.273262439062819ms
  curr LayerNorm backward: 25.28494614805095ms
  apex LayerNorm backward: 13.464738595997915ms
**************************************
Shape = (32768, 8192)
  curr LayerNorm forward: 6.977811902062967ms
  apex LayerNorm forward: 4.115052119130269ms
  curr LayerNorm backward: 25.34099874086678ms
  apex LayerNorm backward: 13.600329003995284ms

```

Performance benchmark for this PR vs apex.FusedLayerNorm on a V100 machine.
```
**************************************
Shape = (128, 2097152)
  curr LayerNorm forward: 7.252584544941783ms
  apex LayerNorm forward: 10.366813436849043ms
  curr LayerNorm backward: 15.568048988003284ms
  apex LayerNorm backward: 20.869979876093566ms
**************************************
Shape = (256, 1048576)
  curr LayerNorm forward: 5.185673736967146ms
  apex LayerNorm forward: 6.3868385690730065ms
  curr LayerNorm backward: 13.942008479032665ms
  apex LayerNorm backward: 15.469660016940907ms
**************************************
Shape = (512, 524288)
  curr LayerNorm forward: 4.672068868065253ms
  apex LayerNorm forward: 4.717993081081659ms
  curr LayerNorm backward: 13.46354596503079ms
  apex LayerNorm backward: 14.04774487693794ms
**************************************
Shape = (1024, 262144)
  curr LayerNorm forward: 4.547273400006816ms
  apex LayerNorm forward: 5.378365494078025ms
  curr LayerNorm backward: 13.425063178874552ms
  apex LayerNorm backward: 14.235145597020164ms
**************************************
Shape = (2048, 131072)
  curr LayerNorm forward: 4.526399010093883ms
  apex LayerNorm forward: 4.775081946980208ms
  curr LayerNorm backward: 13.222738380078226ms
  apex LayerNorm backward: 13.59594238596037ms
**************************************
Shape = (4096, 65536)
  curr LayerNorm forward: 4.28789056581445ms
  apex LayerNorm forward: 4.48913648002781ms
  curr LayerNorm backward: 13.026655421825126ms
  apex LayerNorm backward: 13.57052089786157ms
**************************************
Shape = (8192, 32768)
  curr LayerNorm forward: 4.243518367875367ms
  apex LayerNorm forward: 4.34588153520599ms
  curr LayerNorm backward: 13.140627697808668ms
  apex LayerNorm backward: 13.49891544203274ms
**************************************
Shape = (16384, 16384)
  curr LayerNorm forward: 4.181216162163764ms
  apex LayerNorm forward: 4.268723972840235ms
  curr LayerNorm backward: 13.035593512002379ms
  apex LayerNorm backward: 13.463351831072941ms
**************************************
Shape = (32768, 8192)
  curr LayerNorm forward: 4.097899778978899ms
  apex LayerNorm forward: 4.109480210812762ms
  curr LayerNorm backward: 13.041268918896094ms
  apex LayerNorm backward: 13.586135944118723ms
```

Performance benchmark script:
https://gist.github.com/BIT-silence/cc4221b51519acc5545b2cb8ce6599bc

Differential Revision: D17462420

BIT-silence commented:
linked to Issue https://github.com/pytorch/pytorch/issues/27633

zasdfgbnm commented:
Should we benchmark on `(M, N)` where `M` could be as small as `8` and as large as `4096`, and `N` being `768`, `1024`, `1280`, `2048`? These are the real-world use case I see in https://github.com/huggingface/transformers

zasdfgbnm commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -2321,6 +2321,129 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
+  if (grad_input_mask[0]) {
+    // Let's say
+    // out, mean, rstd = layernorm(input, weight, bias) = (input - mean) * rstd
+    // * weight + bias := f(input, mean, rstd, weight, bias) y = g(out, mean,
```
nit: Should we put `y = ...` in a new line to be more clear?

BIT-silence replied:
> Should we benchmark on `(M, N)` where `M` could be as small as `8` and as large as `4096`, and `N` being `768`, `1024`, `1280`, `2048`? These are the real-world use case I see in https://github.com/huggingface/transformers

The reason I didn't test input with such small size is the overhead in the ops will become comparable large while the result is not very stable.

One of the test with such small input is here https://gist.github.com/BIT-silence/4c762eb6c4db55d58a57a43f492674df
While the timeit result cannot match the profiler result here https://gist.github.com/BIT-silence/27848840dc77f26fc459c8cc5311eb65.
And from the profiler result, we can see that things like empty_like will take a large percentage of time.

Here is the profiler result for the large input https://gist.github.com/BIT-silence/ba01562912bf5815a8059093a87f7c15.
We can see for the large inputs, the numbers from timeit and profiler can matches.

BTW, I found I made a mistake when doing the benchmark that I included the gc.collect time for the 1000 runs, I have changed the benchmark script and update the results above.

BIT-silence replied:
Formatted.

BIT-silence commented:
> Should we benchmark on `(M, N)` where `M` could be as small as `8` and as large as `4096`, and `N` being `768`, `1024`, `1280`, `2048`? These are the real-world use case I see in https://github.com/huggingface/transformers

The reason I didn't test input with such small size is the overhead in the ops will become comparable large while the result is not very stable.

One of the test with such small input is here https://gist.github.com/BIT-silence/4c762eb6c4db55d58a57a43f492674df
While the timeit result cannot match the profiler result here https://gist.github.com/BIT-silence/27848840dc77f26fc459c8cc5311eb65.
And from the profiler result, we can see that things like empty_like will take a large percentage of time.

Here is the profiler result for the large input https://gist.github.com/BIT-silence/ba01562912bf5815a8059093a87f7c15.
We can see for the large inputs, the numbers from timeit and profiler can matches.

BTW, I found I made a mistake when doing the benchmark that I included the gc.collect time for the 1000 runs, I have changed the benchmark script and update the results above.

ngimel commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+    T* dg,
+    T* db) {
+  using T_ACC = acc_type<T, true>;
+  __shared__ T_ACC g_shared[C10_WARP_SIZE][C10_WARP_SIZE + 1];
```
Block dimensions are hardcoded below, and shared memory size is defined through an unrelated macro that happens to be correct. This is fragile.

BIT-silence replied:
Done

ngimel commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+              dbeta_data);
+    } else {
+      const int64_t B = (N + 31) / 32;
+      GammaBetaBackwardCUDAKernel<T><<<B, dim3(32, 16), 0, cuda_stream>>>(
```
Kernel as written would work only if blockDim.x = 2*blockDim.y, which happens to be true, but this limitation should be reflected in the launch parameters (something like `const int threads_y = 16; ...<<<B, dim3(2*threads_y, threads_y), 0, cuda_stream>>>`)

BIT-silence replied:
Done.

ngimel reviewed (changes requested):
Looks good, can you please do a minor cleanup?

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+  const int64_t i = blockIdx.x;
+  T_ACC sum1 = 0;
+  T_ACC sum2 = 0;
+  for (int j = threadIdx.x; j < N; j += blockDim.x) {
```
Done.

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+    const int64_t index = i * N + j;
+    const T gamma_v = gamma == nullptr ? T(1) : gamma[j];
+    const T beta_v = beta == nullptr ? T(0) : beta[j];
+    Y[index] = (X[index] - mean[i]) * rstd[i] * gamma_v + beta_v;
```
Thanks for pointing this out. Done.

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+  sum1 = BlockReduceSum<T_ACC>(sum1);
+  sum2 = BlockReduceSum<T_ACC>(sum2);
+  if (threadIdx.x == 0) {
+    sum1 /= static_cast<T_ACC>(N);
```
Done.

BIT-silence commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,408 @@
+  if (index < M) {
+    const T s = T(1) / static_cast<T>(N);
+    const T a = (db[index] * mean[index] - ds[index]) * rstd[index] *
+        rstd[index] * rstd[index] * s;
```
Done.

houseroad reviewed (commented):
double_backward is missing, do you want to put it into the whitelist, if we are sure nobody serialize them in the save models :-)

ngimel commented on aten/src/ATen/native/cuda/layer_norm_kernel.cu:
```diff
@@ -0,0 +1,442 @@
+  const int lid = threadIdx.x % C10_WARP_SIZE;
+  const int wid = threadIdx.x / C10_WARP_SIZE;
+  val = WarpReduceSum(val);
+  if (lid == 0) {
```
you need another __syncthreads here for 2 BlockReduceSum's called one after another, otherwise it's a race condition. Can you please run the tests with cuda-memcheck --racecheck to make sure that there are no other races?

BIT-silence replied:
Thanks for pointing this out. I changed it to passing shared memory address into BlockReduce then it seems there is no need to do __syncthreads(). I run the tests as well as the benchmark with cuda-memcheck --tool racecheck and there is no racing now.

xiaomengy committed:
Add fused layer norm impl on CUDA in PyTorch (#27634)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27634

Add fused layer norm impl on CUDA in PyTorch

Performance benchmark compare to apex.FusedLayerNorm on a V100 machine.

**************************************
Shape = (128, 2097152)
  curr LayerNorm forward: 7.252584544941783ms
  apex LayerNorm forward: 10.366813436849043ms
  curr LayerNorm backward: 15.568048988003284ms
  apex LayerNorm backward: 20.869979876093566ms
**************************************
Shape = (256, 1048576)
  curr LayerNorm forward: 5.185673736967146ms
  apex LayerNorm forward: 6.3868385690730065ms
  curr LayerNorm backward: 13.942008479032665ms
  apex LayerNorm backward: 15.469660016940907ms
**************************************
Shape = (512, 524288)
  curr LayerNorm forward: 4.672068868065253ms
  apex LayerNorm forward: 4.717993081081659ms
  curr LayerNorm backward: 13.46354596503079ms
  apex LayerNorm backward: 14.04774487693794ms
**************************************
Shape = (1024, 262144)
  curr LayerNorm forward: 4.547273400006816ms
  apex LayerNorm forward: 5.378365494078025ms
  curr LayerNorm backward: 13.425063178874552ms
  apex LayerNorm backward: 14.235145597020164ms
**************************************
Shape = (2048, 131072)
  curr LayerNorm forward: 4.526399010093883ms
  apex LayerNorm forward: 4.775081946980208ms
  curr LayerNorm backward: 13.222738380078226ms
  apex LayerNorm backward: 13.59594238596037ms
**************************************
Shape = (4096, 65536)
  curr LayerNorm forward: 4.28789056581445ms
  apex LayerNorm forward: 4.48913648002781ms
  curr LayerNorm backward: 13.026655421825126ms
  apex LayerNorm backward: 13.57052089786157ms
**************************************
Shape = (8192, 32768)
  curr LayerNorm forward: 4.243518367875367ms
  apex LayerNorm forward: 4.34588153520599ms
  curr LayerNorm backward: 13.140627697808668ms
  apex LayerNorm backward: 13.49891544203274ms
**************************************
Shape = (16384, 16384)
  curr LayerNorm forward: 4.181216162163764ms
  apex LayerNorm forward: 4.268723972840235ms
  curr LayerNorm backward: 13.035593512002379ms
  apex LayerNorm backward: 13.463351831072941ms
**************************************
Shape = (32768, 8192)
  curr LayerNorm forward: 4.097899778978899ms
  apex LayerNorm forward: 4.109480210812762ms
  curr LayerNorm backward: 13.041268918896094ms
  apex LayerNorm backward: 13.586135944118723ms

Test Plan: buck test mode/dev-nosan caffe2/test:nn -- "LayerNorm"

Differential Revision: D17462420

fbshipit-source-id: ab8859272d21b94d9a2849a2255b9b5859ab6e38

gchanan commented:
just a naming thing (and it appears this name probably just comes from apex), but it's strange to refer to an operator as "fused" but not specify what is fused.
